### PR TITLE
add MediaCopy

### PR DESCRIPTION
**Pull request by rawide**
add mediaCopy ddi implementation.
linked libva PR: https://github.com/intel/libva/pull/243
add libva-utils PR: https://github.com/intel/libva-utils/pull/203

### CODE REVIEW

File: media_driver/linux/common/ddi/media_libva.cpp [hunk 1]
```diff
@@ -6002,6 +6002,128 @@ DdiMedia_CopyInternal(
...
+    else if (dst_obj->obj_type == VACopyObjectBuffer)
+    {
+        DDI_CHK_LESS((uint32_t)dst_obj->object.buffer_id, mediaCtx->pBufferHeap->uiAllocatedHeapElements, "Invalid copy dst buf_id", VA_STATUS_ERROR_INVALID_BUFFER);
+        PDDI_MEDIA_BUFFER dst_buffer = DdiMedia_GetBufferFromVABufferID(mediaCtx, dst_obj->object.buffer_id);
+        DDI_CHK_NULL(dst_buffer, "nullptr buffer", VA_STATUS_ERROR_INVALID_BUFFER);
+        DDI_CHK_NULL(dst_buffer->pGmmResourceInfo, "nullptr dst_buffer->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&dst, sizeof(dst));
+        DdiMedia_MediaBufferToMosResource(dst_buffer, &dst);
+    }
```
Comment (dmitryermilov) [inline media_driver/linux/common/ddi/media_libva.cpp:6065]: what about 'else' here?

File: media_driver/linux/common/ddi/media_libva.cpp [hunk 2]
```diff
@@ -6002,6 +6002,128 @@ DdiMedia_CopyInternal(
...
+    else if (src_obj->obj_type == VACopyObjectBuffer)
+    {
+        DDI_CHK_LESS((uint32_t)src_obj->object.buffer_id, mediaCtx->pBufferHeap->uiAllocatedHeapElements, "Invalid copy dst buf_id", VA_STATUS_ERROR_INVALID_BUFFER);
+        PDDI_MEDIA_BUFFER src_buffer = DdiMedia_GetBufferFromVABufferID(mediaCtx, src_obj->object.buffer_id);
+        DDI_CHK_NULL(src_buffer, "nullptr buffer", VA_STATUS_ERROR_INVALID_BUFFER);
+        DDI_CHK_NULL(src_buffer->pGmmResourceInfo, "nullptr src_buffer->pGmmResourceInfo", VA_STATUS_ERROR_INVALID_PARAMETER);
+
+        MOS_ZeroMemory(&src, sizeof(src));
+        DdiMedia_MediaBufferToMosResource(src_buffer, &src);
+    }
```
Comment (dmitryermilov) [inline media_driver/linux/common/ddi/media_libva.cpp:6086]: same

File: media_driver/linux/common/ddi/media_libva.cpp [hunk 3]
```diff
@@ -6002,6 +6002,128 @@ DdiMedia_CopyInternal(
...
+    mosCtx.ppMediaCopyState      = &mediaCtx->pMediaCopyState;
+    mosCtx.pfnMediaCopy          = mediaCtx->pfnMediaCopy;
+    mosCtx.gtSystemInfo          = *mediaCtx->pGtSystemInfo;
+    mosCtx.m_auxTableMgr         = mediaCtx->m_auxTableMgr;
+    mosCtx.pGmmClientContext     = mediaCtx->pGmmClientContext;
+
+    mosCtx.m_osDeviceContext     = mediaCtx->m_osDeviceContext;
+    mosCtx.m_apoMosEnabled       = mediaCtx->m_apoMosEnabled;
+
+    pCpDdiInterface = Create_DdiCpInterface(mosCtx);
```
Comment (dmitryermilov) [inline media_driver/linux/common/ddi/media_libva.cpp:6109]: Create on each copy call? Can we avoid it?

**Comment by dmitryermilov** [comment]
One more thing. This change a relatively simple, it's a proxy to DdiMedia_CopyInternal which seems to do all work. So reviewing this PR only , it's impossible to identify which sub-features of vaCopy the current PR introduce. Under sub-features  I mean VA_COPY_SYNC_BLOCK/VA_COPY_SYNC_NONBLOCK/VA_COPY_MODE_BALANCE/etc , VACopyObjectBuffer/VACopyObjectSurface. Could you describe it in git commit message ?

File: media_driver/linux/common/ddi/media_libva.cpp (hunk 1 above)
Comment (rawide) [reply media_driver/linux/common/ddi/media_libva.cpp:6065]: currently, buffer copy is not supported by vaCopy. buffer copy could be implemented by CPU. carl asked to keep buffer copy as further extension.

File: media_driver/linux/common/ddi/media_libva.cpp (hunk 2 above)
Comment (rawide) [reply media_driver/linux/common/ddi/media_libva.cpp:6086]: as previously.

File: media_driver/linux/common/ddi/media_libva.cpp (hunk 3 above)
Comment (rawide) [reply media_driver/linux/common/ddi/media_libva.cpp:6109]: currently, as driver design. Mos maintain the same MediaCopy as DDI device. and copy is based on Gen specific. mediaCtx will be created during vadpy open, it is also per device. here is just pick up the device existed instance. not created new one.

**Comment by rawide** [comment]
> One more thing. This change a relatively simple, it's a proxy to DdiMedia_CopyInternal which seems to do all work. So reviewing this PR only , it's impossible to identify which sub-features of vaCopy the current PR introduce. Under sub-features I mean VA_COPY_SYNC_BLOCK/VA_COPY_SYNC_NONBLOCK/VA_COPY_MODE_BALANCE/etc , VACopyObjectBuffer/VACopyObjectSurface. Could you describe it in git commit message ?
sample add in libva-utils PR, and VACopy related definition is in libva PR. Do you mean we should add them in this commit again?

File: media_driver/linux/common/ddi/media_libva.cpp (hunk 1 above)
Comment (dmitryermilov) [reply media_driver/linux/common/ddi/media_libva.cpp:6065]: If in VAAPI obj_type gets a new type, how the code above will work? There is no handling of "obj_type  != VACopyObjectSurface && obj_type  != VACopyObjectBuffer"

Comment (rawide) [reply media_driver/linux/common/ddi/media_libva.cpp:6065]: if VAAPI update, suppose media-driver should be updated together. I think. the new type should be defined and verified and then be merged in.

Comment (dmitryermilov) [reply media_driver/linux/common/ddi/media_libva.cpp:6065]: I was walking about error handling. Currently it's absent.

**Comment by dmitryermilov** [comment]
> > One more thing. This change a relatively simple, it's a proxy to DdiMedia_CopyInternal which seems to do all work. So reviewing this PR only , it's impossible to identify which sub-features of vaCopy the current PR introduce. Under sub-features I mean VA_COPY_SYNC_BLOCK/VA_COPY_SYNC_NONBLOCK/VA_COPY_MODE_BALANCE/etc , VACopyObjectBuffer/VACopyObjectSurface. Could you describe it in git commit message ?
> > sample add in libva-utils PR, and VACopy related definition is in libva PR. Do you mean we should add them in this commit again?

If full support of all underlying vaCopy features is appeared by this PR - then I don't think we need additional info in git message. But even above you wrote "currently, buffer copy is not supported by vaCopy". I assume such things better to capture in  git message, together with other limitations if any.

**Comment by rawide** [comment]
> > > One more thing. This change a relatively simple, it's a proxy to DdiMedia_CopyInternal which seems to do all work. So reviewing this PR only , it's impossible to identify which sub-features of vaCopy the current PR introduce. Under sub-features I mean VA_COPY_SYNC_BLOCK/VA_COPY_SYNC_NONBLOCK/VA_COPY_MODE_BALANCE/etc , VACopyObjectBuffer/VACopyObjectSurface. Could you describe it in git commit message ?
> > > sample add in libva-utils PR, and VACopy related definition is in libva PR. Do you mean we should add them in this commit again?
> 
> If full support of all underlying vaCopy features is appeared by this PR - then I don't think we need additional info in git message. But even above you wrote "currently, buffer copy is not supported by vaCopy". I assume such things better to capture in git message, together with other limitations if any.

okay, will add some notification in this git commit message.

File: media_driver/linux/common/ddi/media_libva.cpp (hunk 1 above)
Comment (rawide) [reply media_driver/linux/common/ddi/media_libva.cpp:6065]: okay, add it soon

**Review by dmitryermilov — APPROVED** [review]

**Review by XinfengZhang — APPROVED** [review]
the change is ok for async copy, but still need to check the parameter of copy_sync.

**Pull request merged**